feat(interactive-examples): support CSS examples

caugner opened the pull request:
## Summary

Adds support for CSS examples, especially those with multiple choices.

Part of https://github.com/orgs/mdn/discussions/782.

---

## How did you test this change?

1. Checked out [migrate-interactive-examples](https://github.com/mdn/content/tree/migrate-interactive-examples) content branch.
2. Ran `yarn dev`.
3. Tested especially with these tests:
  - http://localhost:3000/en-US/docs/Web/CSS/background-color#try_it
  - http://localhost:3000/en-US/docs/Web/CSS/text-align#try_it
  - http://localhost:3000/en-US/docs/Web/CSS/color

caugner committed:
refactor(play/editor): extract _defaultExtensions()

caugner committed:
feat(play): add `<play-editor-line>` component

caugner committed:
feat(play): replace line editor by minimal attribute

LeoMcA committed:
wip(ix): prototype css choice logic

caugner committed:
enhance(ix): highlight selected choice

caugner committed:
chore(ix): style CSS Choice header

caugner committed:
chore(ix): sync CSS Choice element design

caugner committed:
chore(ix): add line between choices

caugner committed:
chore(ix): apply CSS Choice base style

caugner committed:
chore(ix): add CSS Choice arrow

caugner committed:
fix(ix): apply CSS Choice height/overflow on first-child

caugner committed:
chore(ix): distribute CSS Choice choices vs output

caugner committed:
feat(ix): mark unsupported CSS choice

caugner committed:
chore(ix): add basic responsive CSS Choice layout

caugner committed:
chore(ix): remove resolved TODO notes

caugner committed:
enhance(ix): select first tab + reset values properly

caugner committed:
enhance(ix): reduce delay for CSS Choice editor

caugner committed:
fix(ix): use correct default height

Previously, values from `.is-js-height` were used.

caugner committed:
fix(ix): select first choice properly

caugner committed:
fix(ix/choices): move padding out of iframe

caugner committed:
fix(ix/choices): unset iframe body font-size

caugner committed:
chore(ix/choice): reduce iframe height

Avoids vertical scrollbar on background-color example.

caugner committed:
chore(ix/choice): reduce space between columns

caugner committed:
fix(ix/choice): scroll on overflow

caugner committed:
chore(play): do not wrap in minimal editor

caugner committed:
fix(ix/choice): avoid border overflow

caugner committed:
fix(ix/choice): use border-bottom on small screens

caugner committed:
chore(ix/choice): hide ::after in vertical layout

caugner committed:
chore(ix/choice): add unsupported border-color to editor

LeoMcA commented on client/src/lit/interactive-example/index.js:
```diff
@@ -95,6 +119,92 @@ export class InteractiveExample extends GleanMixin(LitElement) {
+      contentWindow.postMessage({ typ: "choice", code }, "*");
+    };
+
+    if (contentWindow.document.readyState === "complete") {
```
This line is breaking things outside the dev environment: we can't access `document` cross origin. Instead, we should `postMessage` a "ready" message from inside the runner

LeoMcA replied:
We should expose this from the runner, something like `await runner.ready`

caugner replied:
Fixed via 253cd21a61.

LeoMcA commented on client/src/lit/interactive-example/index.js:
```diff
@@ -95,6 +119,92 @@ export class InteractiveExample extends GleanMixin(LitElement) {
+    // TODO: nicer interface for posting messages than this:
+    const runner = this._runner.value;
+
+    if (!runner) {
+      console.error("No runner");
+      return;
+    }
+
+    // Ensures it has an iframe.
+    await runner.updateComplete;
+
+    const shadowRoot = runner.shadowRoot;
+
+    if (!shadowRoot) {
+      console.error("No shadowRoot");
+      return;
+    }
+
+    const iframe = shadowRoot.querySelector("iframe");
+
+    if (!iframe) {
+      console.error("No iframe");
+      return;
+    }
+
+    const contentWindow = iframe.contentWindow;
+
+    if (!contentWindow) {
+      console.error("No contentWindow");
+      return;
+    }
```
All these checks are very verbose and not really necessary: what we should do is expose a method on `PlayRunner` which exposes `postMessage`

caugner replied:
Fixed via 253cd21a61.

LeoMcA commented on client/src/lit/interactive-example/global.scss:
```diff
@@ -2,9 +2,13 @@
+  height: 375px;
   margin: 1rem 0;
 
+  &[height="js"] {
+    height: 444px;
+  }
```
This doesn't work, there's no `height=js` attribute set anywhere

caugner replied:
Removed in 0090e2932f.

LeoMcA reviewed (commented):
First look:

LeoMcA committed:
wip(ix): migration script now includes section in html, so port styles from bob 1:1

LeoMcA committed:
fix(ix): deal with overflowing editors

LeoMcA committed:
fix(ix): lower specificity of default styles so examples can override

LeoMcA committed:
wip(ix): add transition styling

LeoMcA committed:
Merge remote-tracking branch 'origin/main' into new-interactive-examples-css

LeoMcA committed:
Merge remote-tracking branch 'origin/main' into new-interactive-examples-css

caugner committed:
chore(ix): remove unnecessary console.log

caugner committed:
fix(ix): signal runner ready state via message

Adds the following PlayRunner properties:
- `postMessage()` to post a message when the runner is ready.
- `ready` Promise that resolves when the runner is ready.

caugner commented on client/src/lit/interactive-example/choice.js:
```diff
@@ -0,0 +1,10 @@
+window.addEventListener("message", ({ data }) => {
+  console.log("DATA!", data);
```
Removed in 7ee1ba89b8.

caugner committed:
chore(ix): remove unused "js" height

LeoMcA committed:
wip(ix): simplify some css variables

LeoMcA committed:
wip(ix): don't set live-sample default styles, will move choice.js/css into cloud function after fixing bugs

caugner committed:
chore(play): condense postMessage error handling

caugner committed:
fix(ix): avoid controller reset in chocies

caugner committed:
fix(ix/choices): prefill unsupported mask

caugner committed:
fix(ix/choice): set first choice on initial load

LeoMcA committed:
Merge remote-tracking branch 'origin/main' into new-interactive-examples-css

LeoMcA committed:
fix(ix-choice): scrollbar at bottom and correct background color

LeoMcA committed:
fix(ix-choice): support dark mode

LeoMcA committed:
wip(ix): extract runner defaults into type

LeoMcA committed:
fix(ix): fix csp error for blank.localhost

LeoMcA committed:
fix(ix-choice): keep same height for js demos

LeoMcA committed:
wip(ix): refactor choices/console/tabs into seperate classes

LeoMcA committed:
wip(ix): replace unsupported mask with array

LeoMcA committed:
wip(ix): post ready message after DOM loaded

LeoMcA committed:
wip(ix): move choice js/css into cloud function

LeoMcA committed:
fix(ix): split choice/update event handlers

LeoMcA committed:
wip(ix): small refactor to play-runner using iframe ref

LeoMcA committed:
wip(ix): small fixes to css

LeoMcA committed:
wip(ix): fix check:tsc

LeoMcA committed:
Merge remote-tracking branch 'origin/main' into new-interactive-examples-css

LeoMcA reviewed (commented):
Lovely work, I opted to "show not tell" with my review, and fixed some things/did some refactoring myself, these can be seen with these diffs:

Mostly fixes/additions of things not implemented yet, notably implementing dark mode: https://github.com/mdn/yari/pull/12664/files/b235911713a9e2f299494bd52d479a471a44b5f5..46cc5847f2dd5d6722f1db32dd02ba76c1041b4b

A pretty major refactor which I think you'd mentioned might be good to do, separating out the choice logic (and tabbed, and console) - achieved it in a perhaps unconventional way (using mixins), but it works nicely: https://github.com/mdn/yari/pull/12664/commits/97fd3a37188151466d70a64c779bcfe3560a89f2

Mostly code review changes, notably replacing the bitmask with an array and splitting the click/update events: https://github.com/mdn/yari/pull/12664/files/97fd3a37188151466d70a64c779bcfe3560a89f2..8bcafda92f3e054574dd5eba29f51620bfbc8299

LeoMcA committed:
Merge remote-tracking branch 'origin/main' into new-interactive-examples-css

fiji-flo reviewed (approved):
:+1: for your changes @LeoMcA
